File: production-data-visualization-backend-advanced/backend/src/controllers/dataSourceController.js
const dataSourceService = require('../services/dataSourceService');
const catchAsync = require('../utils/catchAsync');
const logger = require('../utils/logger');

exports.createDataSource = catchAsync(async (req, res) => {
  const dataSource = await dataSourceService.createDataSource(req.body, req.user.id);
  res.status(201).json({
    status: 'success',
    data: { dataSource },
  });
  logger.info(`Data source created: ${dataSource.name} by user ${req.user.id}`);
});

exports.getAllDataSources = catchAsync(async (req, res) => {
  const dataSources = await dataSourceService.getAllDataSources(req.user.id, req.user.role);
  res.status(200).json({
    status: 'success',
    results: dataSources.length,
    data: { dataSources },
  });
  logger.debug(`Fetched all data sources for user ${req.user.id}`);
});

exports.getDataSource = catchAsync(async (req, res) => {
  const dataSource = await dataSourceService.getDataSourceById(req.params.id, req.user.id, req.user.role);
  res.status(200).json({
    status: 'success',
    data: { dataSource },
  });
  logger.debug(`Fetched data source: ${req.params.id}`);
});

exports.updateDataSource = catchAsync(async (req, res) => {
  const dataSource = await dataSourceService.updateDataSource(req.params.id, req.body, req.user.id, req.user.role);
  res.status(200).json({
    status: 'success',
    data: { dataSource },
  });
  logger.info(`Updated data source: ${dataSource.name}`);
});

exports.deleteDataSource = catchAsync(async (req, res) => {
  await dataSourceService.deleteDataSource(req.params.id, req.user.id, req.user.role);
  res.status(204).json({
    status: 'success',
    data: null,
  });
  logger.warn(`Deleted data source: ${req.params.id}`);
});

exports.getDataSourceData = catchAsync(async (req, res, next) => {
  const data = await dataSourceService.getDataSourceData(req.params.id, req.user.id, req.user.role);
  res.status(200).json({
    status: 'success',
    data: { data },
  });
  logger.debug(`Fetched data for data source: ${req.params.id}`);
});

File: scalable-data-visualization-solution-optimized/backend/src/main/java/com/datavizpro/datavizpro/datasource/DataSourceController.java
package com.datavizpro.datavizpro.datasource;

import com.datavizpro.datavizpro.shared.dtos.ApiResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/data-sources")
@RequiredArgsConstructor
@SecurityRequirement(name = "bearerAuth") // For Swagger UI
@Tag(name = "Data Sources", description = "API for managing data sources")
public class DataSourceController {

    private final DataSourceService dataSourceService;

    @Operation(summary = "Create a new data source")
    @PostMapping
    @PreAuthorize("hasRole('ADMIN') or hasRole('USER')")
    public ResponseEntity<ApiResponse<DataSourceDto>> createDataSource(@Valid @RequestBody DataSourceCreateUpdateDto dto, @AuthenticationPrincipal UserDetails currentUser) {
        DataSourceDto createdDataSource = dataSourceService.createDataSource(dto, currentUser.getUsername());
        return new ResponseEntity<>(ApiResponse.success("Data source created successfully", createdDataSource), HttpStatus.CREATED);
    }

    @Operation(summary = "Get all data sources for the current user (or all if ADMIN)")
    @GetMapping
    @PreAuthorize("hasRole('ADMIN') or hasRole('USER') or hasRole('VIEWER')")
    public ResponseEntity<ApiResponse<List<DataSourceDto>>> getAllDataSources(@AuthenticationPrincipal UserDetails currentUser) {
        List<DataSourceDto> dataSources = dataSourceService.getAllDataSourcesForUser(currentUser.getUsername());
        return ResponseEntity.ok(ApiResponse.success("Data sources retrieved successfully", dataSources));
    }

    @Operation(summary = "Get a data source by ID")
    @GetMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or @dataSourceService.isOwner(#id, authentication.name)")
    public ResponseEntity<ApiResponse<DataSourceDto>> getDataSourceById(@PathVariable Long id) {
        DataSourceDto dataSource = dataSourceService.getDataSourceById(id);
        return ResponseEntity.ok(ApiResponse.success("Data source retrieved successfully", dataSource));
    }

    @Operation(summary = "Update an existing data source")
    @PutMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or @dataSourceService.isOwner(#id, authentication.name)")
    public ResponseEntity<ApiResponse<DataSourceDto>> updateDataSource(@PathVariable Long id, @Valid @RequestBody DataSourceCreateUpdateDto dto) {
        DataSourceDto updatedDataSource = dataSourceService.updateDataSource(id, dto);
        return ResponseEntity.ok(ApiResponse.success("Data source updated successfully", updatedDataSource));
    }

    @Operation(summary = "Delete a data source by ID")
    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or @dataSourceService.isOwner(#id, authentication.name)")
    public ResponseEntity<ApiResponse<Void>> deleteDataSource(@PathVariable Long id) {
        dataSourceService.deleteDataSource(id);
        return ResponseEntity.ok(ApiResponse.success("Data source deleted successfully"));
    }

    @Operation(summary = "Test connection to a data source")
    @PostMapping("/test-connection")
    @PreAuthorize("hasRole('ADMIN') or hasRole('USER')")
    public ResponseEntity<ApiResponse<String>> testDataSourceConnection(@Valid @RequestBody DataSourceCreateUpdateDto dto) {
        // In a real scenario, this would involve connecting and executing a simple query
        boolean success = dataSourceService.testConnection(dto);
        if (success) {
            return ResponseEntity.ok(ApiResponse.success("Connection successful", "Connection to data source established."));
        } else {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(ApiResponse.error("Connection failed", "Could not connect to data source with provided details."));
        }
    }
}